### DISCUSSION TITLE
HID Device (Keyboard) Boot Protocol for use in PC BIOS/GRUB etc

### DISCUSSION THREAD

**Discussion by ndt1896** [hathach/tinyusb#1003] [Q&A] [answered]
I am looking to try to implement a HID keyboard that could work in a computer BIOS. There appears to be some support coming for HID Device Boot Protocol but not implemented yet(?).  

What would be the high level steps to make a HID Keyboard device that could work in a PC BIOS?

My guess would be: 

-> Try to set the protocol to BOOT PROTOCOL in the code (but where?)
-> Modify the device descriptors to match a keyboard known to work in a BIOS.

Much appreciated for any help getting on the right direction to try this.

**Comment by hathach** [comment]
you need to use the boot protocol in hid descriptor, may need an extra/dedicated interface for it (though I don't think it is necessary). For reading info is discussed here https://github.com/adafruit/circuitpython/issues/1136 . I also plan to add/update hid example with boot protocol, it is not difficult at all but haven't got time to do so. Pay attention to future PRs for it.

**Comment by hathach** [comment, marked as answer]
I just add an PR #1025 to add boot interface example, however I don't have the legacy bios PC to test with. Would you mind trying it out /

**Reply by ndt1896** [reply]
Thank you for the exciting update!  I'll test on some legacy BIOS PC's and update findings here.

**Reply by ndt1896** [reply]
Today I tested on one Lenovo (consumer) desktop and one HP (enterprise) Server.  The tests on real hardware worked perfectly!  Thank you so much.  Keyboard and Mouse worked in BIOS, Grub, and iPXE menus without issue.
